### ISSUE TITLE
Windows AMD drivers are officially too shitty to run Taisei

### ISSUE THREAD

**Issue by Akaricchi** [taisei-project/taisei#146]
Confirmed for Radeon R9 M275 and Radeon HD 5800 Series. The stage1 water shader doesn't compile, error message makes no sense:

```
ERROR: 0:19: error(#53) Built-in function parameter type check fail: an operation which requires matrix parameters is used on a non matrix type!
ERROR: 0:19: error(#53) Built-in function parameter type check fail: an operation which requires matrix parameters is used on a non matrix type!
```

3D models are broken in the HD5800 case, R9 M275 unconfirmed.

**Comment by Akaricchi** [comment]
I don't know if I can work around this. The only AMD card I have is an RX 580 on my main system - which doesn't run Windows and doesn't have these problems. I guess I'll have to install Windows on it and see if I can reproduce the issue there.

However, I'm leaning towards basically ignoring this, and instead focusing on finishing the GLES 3.0 backend for use with ANGLE. Windows drivers are generally less retarded when it comes to D3D.

If anyone has a clue as to what exactly the AMD driver hates about [this shader](../blob/master/resources/shader/stage1_water.frag.glsl), though, hit me up.

**Comment by Akaricchi** [comment]
Update on this one: apparently the backgrounds issue was caused by us still using client-side index buffers, which is not supported in OpenGL 3.3. The reason why it took me so long to figure it out is, while they are *not supposed* to work in a Core context, in most implementations they, apparently, do. AMD's windows driver seems to be the only one that actually follows the spec here (credit where it's due). I've implemented IBOs in a6d0ab565fba7bb31e3cf1b381d9054585bac278. That should fix it, though I have not tested it on windows yet. It is also more efficient than passing the index arrays around on every draw, obviously.

I still have no clue about the shader error, however.
